pkg/flowaggregator/flowaggregator.go

```diff
@@ -15,25 +15,250 @@
+		"reversePacketTotalCount",
+		"reverseOctetTotalCount",
+		"reversePacketDeltaCount",
+		"reverseOctetDeltaCount",
+	}
+	antreaInfoElements = []string{
+		"sourcePodName",
+		"sourcePodNamespace",
+		"sourceNodeName",
+		"destinationPodName",
+		"destinationPodNamespace",
+		"destinationNodeName",
+		"destinationServicePort",
+		"destinationServicePortName",
+		"ingressNetworkPolicyName",
+		"ingressNetworkPolicyNamespace",
+		"egressNetworkPolicyName",
+		"egressNetworkPolicyNamespace",
+		"destinationClusterIPv4",
+	}
+	aggregatorElements = []string{
+		"originalExporterIPv4Address",
+		"originalObservationDomainId",
+	}
+	aggregationFields = []string{
+		"destinationPodName",
+		"destinationPodNamespace",
+		"destinationNodeName",
+		"destinationClusterIPv4",
+		"destinationServicePortName",
+		"ingressNetworkPolicyName",
+		"ingressNetworkPolicyNamespace",
+	}
+	aggregationWorkerNum = 2
 )
 
 type flowAggregator struct {
-	flowCollectorAddr net.Addr
-	exportInterval    time.Duration
+	externalFlowCollectorAddr net.Addr
+	collectingProtocol        string
+	collectingProcess         ipfix.IPFIXCollectingProcess
+	aggregationProcess        ipfix.IPFIXAggregationProcess
+	exportInterval            time.Duration
+	exportingProcess          ipfix.IPFIXExportingProcess
+	templateID                uint16
+	registry                  ipfix.IPFIXRegistry
 }
 
-func InitFlowAggregator(flowCollectorAddr net.Addr, exportInterval time.Duration) *flowAggregator {
+func NewFlowAggregator(externalFlowCollectorAddr net.Addr, exportInterval time.Duration, collectingProtocol string) *flowAggregator {
+	registry := ipfix.NewIPFIXRegistry()
 	registry.LoadRegistry()
-	return &flowAggregator{
-		flowCollectorAddr,
+	fa := &flowAggregator{
+		externalFlowCollectorAddr,
+		collectingProtocol,
+		nil,
+		nil,
 		exportInterval,
+		nil,
+		0,
+		registry,
+	}
+	return fa
+}
+
+func genObservationID() (uint32, error) {
+	name, err := env.GetNodeName()
+	if err != nil {
+		return 0, err
+	}
+	h := fnv.New32()
+	h.Write([]byte(name))
+	return h.Sum32(), nil
+}
+
+func (fa *flowAggregator) InitCollectingProcess() error {
+	var collectAddr net.Addr
+	var err error
+	if fa.collectingProtocol == "tcp" {
+		collectAddr, _ = net.ResolveTCPAddr("tcp", "0.0.0.0:4739")
+		fa.collectingProcess, err = ipfix.NewIPFIXCollectingProcess(collectAddr, 65535, 0)
+	} else {
+		collectAddr, _ = net.ResolveUDPAddr("udp", "0.0.0.0:4739")
+		fa.collectingProcess, err = ipfix.NewIPFIXCollectingProcess(collectAddr, 1024, 0)
 	}
+	return err
+}
+
+func (fa *flowAggregator) InitAggregationProcess() error {
+	var err error
+	fa.aggregationProcess, err = ipfix.NewIPFIXAggregationProcess(fa.collectingProcess.GetMsgChan(), aggregationWorkerNum, aggregationFields)
+	return err
+}
+
+func (fa *flowAggregator) initExportingProcess() error {
+	obsID, err := genObservationID()
+	if err != nil {
+		return fmt.Errorf("cannot generate obsID for IPFIX ipfixexport: %v", err)
```

Comment:
not a great error message:
 - spell out observation ID
 - what is ipfixexport?

Reply:
Thanks, addressed.